src/machine.rs

```diff
@@ -1243,7 +1243,23 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
             // foreign function
             // Any needed call to `goto_block` will be performed by `emulate_foreign_item`.
             let args = ecx.copy_fn_args(args); // FIXME: Should `InPlace` arguments be reset to uninit?
-            let link_name = Symbol::intern(ecx.tcx.symbol_name(instance).name);
+
+            let link_name = if ecx.tcx.sess.target.is_like_wasm
+                && let Some(module) =
+                    ecx.tcx.wasm_import_module_map(instance.def_id().krate).get(&instance.def_id())
```

Comment:
I hope this code doesn't have to live in `machine.rs`.

If possible, please move it inside the `wasi` folder.
If that doesn't work, at least push it down to inside `emulate_foreign_item`.

Reply:
`emulate_foreign_item` is also called from `emulate_dyn_sym` which doesn't have access to the `Instance` as `dlsym` never gets passed one. I guess I could pass an `Option<Instance>` to `emulate_foreign_item` if you prefer that.